sklearn/inspection/_metric_threshold_curve.py

```diff
@@ -0,0 +1,177 @@
+    """
+    # Check to make sure y_true is valid.
+    y_type = type_of_target(y_true, input_name="y_true")
+    if not (y_type == "binary" or (y_type == "multiclass" and pos_label is not None)):
+        raise ValueError("{0} format is not supported".format(y_type))
+
+    check_consistent_length(y_true, y_score, sample_weight)
+    y_true = column_or_1d(y_true)
+    y_score = column_or_1d(y_score)
+    assert_all_finite(y_true)
+    assert_all_finite(y_score)
+
+    # Filter out zero-weighted samples, as they should not impact the result.
+    if sample_weight is not None:
+        sample_weight = column_or_1d(sample_weight)
+        sample_weight = _check_sample_weight(sample_weight, y_true)
+        nonzero_weight_mask = sample_weight != 0
+        y_true = y_true[nonzero_weight_mask]
+        y_score = y_score[nonzero_weight_mask]
+        sample_weight = sample_weight[nonzero_weight_mask]
+
+    pos_label = _check_pos_label_consistency(pos_label, y_true)
+
+    # Make y_true a boolean vector.
+    y_true = y_true == pos_label
+
+    # Sort scores and corresponding truth values.
+    desc_score_indices = np.argsort(y_score, kind="mergesort")[::-1]
+    y_score = y_score[desc_score_indices]
+    y_true = y_true[desc_score_indices]
+    if sample_weight is not None:
+        sample_weight = sample_weight[desc_score_indices]
+
+    # Logic to see if we need to use all possible thresholds (distinct values).
+    all_thresholds = False
+    if threshold_grid is None:
+        all_thresholds = True
+    elif isinstance(threshold_grid, int):
+        if len(set(y_score)) < threshold_grid:
+            all_thresholds = True
+
+    if all_thresholds:
+        # y_score typically has many tied values. Here we extract
+        # the indices associated with the distinct values. We also
+        # concatenate a value for the end of the curve.
+        distinct_value_indices = np.where(np.diff(y_score))[0]
+        threshold_idxs = np.r_[distinct_value_indices, y_true.size - 1]
+        thresholds = y_score[threshold_idxs[::-1]]
+    elif isinstance(threshold_grid, int):
+        # It takes representative score points to calculate the metric
+        # with these thresholds.
+        thresholds = np.percentile(
+            list(set(y_score)), np.linspace(0, 100, threshold_grid)
+        )
+    else:
+        # If threshold_grid is an array then run some checks and sort
+        # it for consistency.
+        threshold_grid = column_or_1d(threshold_grid)
+        assert_all_finite(threshold_grid)
+        thresholds = np.sort(threshold_grid)
+
+    # For each threshold calculates the metric.
+    metric_values = []
+    for threshold in thresholds:
+        preds_threshold = (y_score > threshold).astype(int)
+        metric_values.append(
+            score_func(y_true, preds_threshold, sample_weight=sample_weight)
+        )
+    # TODO: should we multithread the metric calculations?
+
+    return np.array(metric_values), thresholds
```

Comment:
All this code is already implemented by the `_score` function of the `sklearn.model_selection._classification_threshold._CurveScorer` class.

I think that we should leverage this code by creating this scorer. We probably need to dissociate getting `y_score` from the scoring itself such that here we only call the scoring part.

Reply:
So now, it makes sense to move the `_CurveScorer` in `metrics`.

Reply:
> So now, it makes sense to move the `_CurveScorer` in `metrics`.

Do you want me to do this in this PR or create a separate one?

Reply:
It would be better to be in a separate PR.

Reply:
> We probably need to dissociate getting y_score from the scoring itself such that here we only call the scoring part.

I think it will make sense to do this on this PR after we move the code to the proper place. Do you agree?